DREAM/Optimization/tsgParticleSwarm.cpp

```diff
@@ -0,0 +1,146 @@
+    for (int i=0; i<num_particles; i++) {
+        for (int j=0; j<num_dimensions; j++) {
+            double range = std::fabs(box_upper[j] - box_lower[j]);
+            int idx = i * num_dimensions + j;
+            particle_positions[idx] = range * get_random01() + box_lower[j];
+            particle_velocities[idx] = 2 * range * get_random01() - range;
+        }
+    }
+    initialized = true;
+}
+
+
+void ParticleSwarm(ObjectiveFunction f, int max_iterations, TasDREAM::DreamDomain inside, ParticleSwarmState &state,
+                   double inertia_weight, double cognitive_coeff, double social_coeff, std::function<double(void)> get_random01) {
+
+    // Only run the algorithm on properly initialized states.
+    if (!state.initialized) {
+        throw std::runtime_error("Particles have not been initialized in the input state object");
+    }
+
+    // Initialize helper variables and functions.
+    size_t num_dimensions = (size_t) state.getNumDimensions();
+    size_t num_particles = (size_t) state.getNumParticles();
+    std::vector<double> &particle_positions = state.getParticlePositionsRef();
+    std::vector<double> &particle_velocities = state.getParticleVelocitiesRef();
+    std::vector<double> &best_particle_positions = state.getBestParticlePositionsRef();
+    std::vector<double> &best_position = state.getBestPositionRef();
+    std::vector<double> prev_best_vals(num_particles), cur_best_vals(num_particles);
+    std::vector<bool> prev_is_inside(num_particles), cur_is_inside(num_particles);
+    double best_fval = std::numeric_limits<double>::max();
+    ObjectiveFunctionConstrained f_constrained = makeObjectiveFunctionConstrained(num_dimensions, f, inside);
+
+    // Create a lambda that updates the best particle and swarm positions based on function values and domain information.
+    auto update_best_positions = [&](const std::vector<double> prev_vals, const std::vector<double> cur_vals,
+                                     const std::vector<bool> prev_domain, const std::vector<bool> cur_domain)->void {
+        for (size_t i=0; i<num_particles; i++) {
+            bool is_smaller = (cur_vals[i] < prev_vals[i]) || !prev_domain[i];
+            if (cur_domain[i] && is_smaller) {
+                std::copy(particle_positions.begin() + i * num_dimensions, particle_positions.begin() + (i + 1) * num_dimensions,
+                          best_particle_positions.begin() + i * num_dimensions);
+                if (cur_vals[i] < best_fval) {
+                    std::copy(particle_positions.begin() + i * num_dimensions, particle_positions.begin() + (i + 1) * num_dimensions,
+                              best_position.begin());
+                    best_fval = cur_vals[i];
+                }
+            }
+        }
+    };
+
+    // Initialize the best particle and swarm function values, while updating the input state.
+    f_constrained(best_particle_positions, prev_best_vals, prev_is_inside);
```

Comment:
This should not be needed, if the user has provided an initialization of `best_particle_position` then the user should also provide the values. Also, the values would be know if those come from an earlier call to the particle swam candicate.

Reply:
This part is a bit tricky because we are passing `inside()` as an input. 

More specifically, this is because a particular best particle value should be finite only if the corresponding point is inside the domain prescribed (which will depend on `inside()`), and there is no guarantee that the user will always pass in the same `inside()` function between calls.

Reply:
We are updating the state to hold cached values for the `inside()`.